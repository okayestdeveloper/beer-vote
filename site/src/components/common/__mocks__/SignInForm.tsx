import React from 'react';

const SignInForm: React.FC<any> = () => {
  return <form>sign in form</form>;
};

export default SignInForm;
