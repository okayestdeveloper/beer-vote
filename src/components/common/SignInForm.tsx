import React, {
  useState,
  FocusEvent,
  FormEvent,
  ChangeEvent,
  useEffect,
} from 'react';
import { makeStyles, Theme, createStyles } from '@material-ui/core';
import Typography from '@material-ui/core/Typography';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import Box from '@material-ui/core/Box';

import { IProfile } from './../../hooks/useProfile';
import { isValidEmail } from './../../shared/utils';

// todo: tests

export interface ISignInProps {
  signIn: (profile: IProfile) => void;
  visible: boolean;
}

interface ISignInFormState extends IProfile {
  emailError: boolean;
  nameError: boolean;
}

const useStyles = makeStyles((theme: Theme) => {
  const formControl = {
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(1),
  };
  return createStyles({
    signInFormContainer: {
      margin: theme.spacing(1),
      padding: theme.spacing(1),
      minWidth: '25rem',
    },
    signInForm: {
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'flex-start',
    },
    formControl,
    submitButton: {
      ...formControl,
      alignSelf: 'flex-end',
    },
  });
});

const SignInForm: React.FC<ISignInProps> = ({ signIn, visible }) => {
  const [formState, setFormState] = useState<ISignInFormState>({
    email: '',
    name: '',
    emailError: false,
    nameError: false,
  });

  const classes = useStyles();

  // note: this is my hacky workaround for an "on show"
  useEffect(() => {
    // todo: handle resetting the form when showing
    // todo: handle focusing the email input when showing
    console.log(`SignInForm visibility: ${visible ? 'visible' : 'hidden'}`);
  }, [visible]);

  const handleSignIn = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    signIn(formState);
  };

  // todo: create email and name components and have them validate themselves

  const validate = (event: FocusEvent<HTMLInputElement>) => {
    switch (event.target.id) {
      case 'email':
        if (!event.target.value || !isValidEmail(event.target.value)) {
          setFormState({ ...formState, emailError: true });
        } else {
          setFormState({ ...formState, emailError: false });
        }
        break;
      case 'name':
        if (!event.target.value) {
          setFormState({ ...formState, nameError: true });
        } else {
          setFormState({ ...formState, nameError: false });
        }
        break;
    }
  };

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    if (event?.target?.id) {
      setFormState({ ...formState, [event.target.id]: event.target.value });
    }
  };

  return (
    <Box className={classes.signInFormContainer}>
      <Typography variant="body1">Sign In</Typography>
      <form onSubmit={handleSignIn} className={classes.signInForm}>
        <TextField
          label="Email"
          value={formState.email}
          onChange={handleChange}
          id="email"
          type="email"
          required
          className={classes.formControl}
          onBlur={validate}
          error={formState.emailError}
          helperText={
            formState.emailError ? 'Please enter a valid email address.' : ''
          }
        />
        <TextField
          label="Name"
          value={name}
          onChange={handleChange}
          id="name"
          required
          className={classes.formControl}
          onBlur={validate}
          error={formState.nameError}
          helperText={formState.nameError ? 'Please enter your name.' : ''}
        />
        <Button
          variant="contained"
          color="primary"
          type="submit"
          className={classes.submitButton}
        >
          Sign In
        </Button>
      </form>
    </Box>
  );
};

export default SignInForm;
