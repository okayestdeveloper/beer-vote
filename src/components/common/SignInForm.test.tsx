import React from 'react';
import renderer from 'react-test-renderer';
import {
  render,
  cleanup,
  fireEvent,
  RenderResult,
} from '@testing-library/react';

import SignInForm, { ISignInProps } from './SignInForm';

function renderSignInForm(args?: ISignInProps): RenderResult {
  const defaultProps: ISignInProps = {
    onChange: jest.fn(),
  };

  const props = { ...defaultProps, ...args };
  return render(<SignInForm {...props} />);
}

describe(`SignInForm`, () => {
  describe(`snapshots`, () => {
    it(`should render a form with email and name controls`, () => {
      const tree = renderer.create(<SignInForm onChange={jest.fn()} />);
      expect(tree).toMatchSnapshot();
    });
  });

  describe(`form changes`, () => {
    let onChange: jest.Mock;
    let emailEl: HTMLInputElement;
    let nameEl: HTMLInputElement;

    beforeEach(() => {
      onChange = jest.fn();
      const { getByPlaceholderText } = renderSignInForm({ onChange });
      nameEl = getByPlaceholderText('Name') as HTMLInputElement;
      emailEl = getByPlaceholderText('Email') as HTMLInputElement;
    });

    afterEach(cleanup);

    it(`should call onChange with new name when name changes`, () => {
      const value = 'Frank';
      fireEvent.change(nameEl, { target: { value } });

      expect(onChange).toHaveBeenCalledWith(
        expect.objectContaining({ name: value }),
      );
    });

    it(`should call onChange with new email when email changes`, () => {
      const value = 'a@b.com';
      fireEvent.change(emailEl, { target: { value } });

      expect(onChange).toHaveBeenCalledWith(
        expect.objectContaining({ email: value }),
      );
    });

    it(`should call onChange with valid:false if form is invalid`, () => {
      fireEvent.change(emailEl, { target: { value: 'not good' } });
      fireEvent.blur(emailEl);

      expect(onChange).toHaveBeenCalledWith(
        expect.objectContaining({ valid: false }),
      );
    });
  });
});
