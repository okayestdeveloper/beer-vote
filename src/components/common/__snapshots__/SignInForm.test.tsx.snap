// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInForm snapshots should render a form with email and name controls 1`] = `
Array [
  <p
    className="MuiTypography-root MuiTypography-body1"
  >
    Sign In
  </p>,
  <form
    className="makeStyles-signInForm-1"
  >
    <div
      className="MuiFormControl-root MuiTextField-root makeStyles-formControl-2"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
      >
        Email
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          name="Email"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Email"
          required={false}
          type="email"
          value=""
        />
      </div>
      
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root makeStyles-formControl-2"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
      >
        Name
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          name="Name"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Name"
          required={false}
          type="text"
          value=""
        />
      </div>
      
    </div>
  </form>,
]
`;
